### DISCUSSION TITLE
UpdateModel from server to client

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-glsp/glsp#937] [Q&A]
Hi,
    I change modelstate value in my glsp-server, I want to update client.  I look at code and find ModelSubmissionHandler. my code as follow
![image](https://user-images.githubusercontent.com/4537809/221618110-28446d47-3a38-401a-96eb-f5a875469cbc.png)

It seem that UpdateModelCommand is used to deal with UpdateModelAction in client.  but it is unuseful. 
![image](https://user-images.githubusercontent.com/4537809/221618203-48aab382-328b-42cc-90b5-6f72a0f72a00.png)
![image](https://user-images.githubusercontent.com/4537809/221617851-94a0d8b9-9d9b-4abb-9b4a-b91ebd80cc23.png)
 any mistake do I have?

**Comment by planger** [comment]
Hi,

your approach to use the `ModelSubmissionHandler` -- similar to how it is done in the [OperationActionHandler](https://github.com/eclipse-glsp/glsp-server-node/blob/main/packages/server/src/common/operations/operation-action-handler.ts) -- is correct and your code looks fine at a first glance, except maybe for the use of `submitModelDirectly()` instead of `submitModel()`. Is there a particular reason for this choice? Calling `submitModelDirectly()` would bypass any bounds computation and client-side layouting and is usually only used by the `ComputedBoundsActionHandler` after all of that client-side layouting already happened.

So I'd try if `submitModel()` instead of `submitModelDirectly()` improves the situation. Alternatively, you could try to dispatch an `UpdateModelAction` with a new GModel (which you can create from your model state with a `@inject(GModelFactory) protected modelFactory: GModelFactory)`) directly. We do something similar also in the [workflow demo](https://github.com/eclipse-glsp/glsp-server/blob/advanced-demo/examples/org.eclipse.glsp.example.workflow/src/org/eclipse/glsp/example/workflow/handler/FilterActionHandler.java).

If all of that doesn't help, I'm afraid, I don't have a good suggestion except for debugging into the client code to see what's wrong.

**Reply by wss29** [reply]
Hi,
   it is ok to use submitModel, I did not understand submitModelDirectly clearly last week. but there is a slight conflict about updating model.
![BCB814C3-2A1F-4E6B-8891-16C5210FD4BE](https://user-images.githubusercontent.com/4537809/222043916-f102b1cd-e20a-4559-9574-a0e3ff3003f6.GIF)

![image](https://user-images.githubusercontent.com/4537809/222043283-02d00926-bfdb-4471-a539-7c4b22d989dc.png)
there are shakes between positions when I drag on the client. I guess I drag the node new position update from client to server and meanwhile model update from server to client， Is there any solution for this situation?

**Reply by planger** [reply]
For this purpose we have the [`FeedbackActionDispatcher`](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/tool-feedback/feedback-action-dispatcher.ts) which is supposed to reapply visual feedback on top of the updated model, if an `UpdateModelAction` happens to come in while we show feedback.
It looks like we don't use this mechanism though in the [change bounds tool feedback](https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/tool-feedback/change-bounds-tool-feedback.ts). See also:
https://github.com/eclipse-glsp/glsp-client/blob/master/packages/client/src/features/tool-feedback/change-bounds-tool-feedback.ts#L152

So I think this is actually a bug and we'd need to change the `FeedbackMoveMouseListener` to make it use the `FeedbackActionDispatcher`. Would you be able to create a bug report and open a PR on that?

I think periodic updates sent from the server is a use case that we never really had so far, thus this issue was never observed.

Thank you!

**Reply by wss29** [reply]
Hi,
    I create a bug report at https://github.com/eclipse-glsp/glsp/issues/941#issue-1606329001
